### PR TITLE
A0-3390: Proposals with headers

### PR DESCRIPTION
**Pull request by timorleph**
# Description

Proposals now include headers.

## Type of change

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- I have made corresponding changes to the existing documentation

### CODE REVIEW

File: finality-aleph/src/data_io/data_provider.rs [hunk 1]
```diff
@@ -77,17 +84,25 @@ where
...
+        let head_hash = match branch.next() {
+            Some(hash) => hash,
+            None => return Ok(None),
+        };
+        let head = match client.header(head_hash) {
+            Ok(Some(header)) => header,
+            _ => return Err(MissingHeader),
+        };
+        let mut tail: Vec<_> = branch.collect();
+        tail.reverse();
```
Comment (kostekIV) [inline finality-aleph/src/data_io/data_provider.rs:97]:
```suggestion
        let tail: Vec<_> = branch.rev().collect()
```

**Review by kostekIV — APPROVED** [review]
LGTM

**Review by ggawryal — APPROVED** [review]
LGTM

File: finality-aleph/src/abft/current/mod.rs [hunk 2]
```diff
@@ -28,23 +29,31 @@ use crate::{
...
     >,
-    data_provider: impl current_aleph_bft::DataProvider<AlephData> + Send + 'static,
-    ordered_data_interpreter: OrderedDataInterpreter<SubstrateChainInfoProvider<B, C>>,
     backup: ABFTBackup,
 ) -> Task
 where
     B: Block<Hash = BlockHash>,
-    B::Header: Header<Number = BlockNumber>,
+    B::Header:
+        Header<Number = BlockNumber> + UnverifiedHeader + BlockHeader<Unverified = B::Header>,
```
[the remark this answers is not in the document]
Comment (timorleph) [reply finality-aleph/src/abft/current/mod.rs:53]: It doesn't really make internal sense there, these can theoretically be different types underneath (even if the implementation we use is the same type).

These requirements will look much less convoluted after A0-2225 gets done, but until then... eh.

**Pull request merged**